core/node/rpc/sync/client/syncer_set.go

```diff
@@ -434,57 +484,162 @@ func (ss *SyncerSet) modify(ctx context.Context, req ModifyRequest) error {
 	if err != nil {
 		return err
 	}
 
-	delete(ss.streamID2Syncer, streamID)
+	ss.streamID2Syncer.Delete(streamID)
 	if syncerStopped {
-		delete(ss.syncers, syncer.Address())
+		ss.syncers.Delete(syncer.Address())
 	}
 
 	return nil
 }
 
-func (ss *SyncerSet) rmStream(streamID StreamId) {
-	ss.muSyncers.Lock()
-	delete(ss.streamID2Syncer, streamID)
-	ss.muSyncers.Unlock()
+// startSyncingByCookie resyncs the given stream from the given cookie.
+// Gets the stream from either local or remote node and sends stream updates to the given subscription.
+// This function is called when a new subscription for the given stream is created with a specific cookie but
+// the stream is already part of the sync operation and has the latest state so this function loads the stream
+// since the given cookie and sends the updates to the subscription.
+// The given cookie must contain the node address to use.
+func (ss *SyncerSet) startSyncingByCookie(
+	ctx context.Context,
+	syncID string,
+	cookie *SyncCookie,
+) error {
+	streamID := StreamId(cookie.GetStreamId())
+
+	// Check if the given sync is already subscribed to the given stream.
+	if ss.router.isSubscribed(syncID, streamID) {
+		return nil
+	}
+
+	if cookie.GetMinipoolGen() > 0 {
+		// Get subscription by the given sync ID
+		sub := ss.router.getSubscription(syncID)
+		if sub == nil {
+			return RiverError(Err_NOT_FOUND, "Subscription not found for the given stream").
+				Tags("syncId", syncID, "streamID", streamID).Func("startSyncingByCookie")
+		}
+
+		// Get stream from either local or remote node
+		var sc *StreamAndCookie
+		addr := common.BytesToAddress(cookie.GetNodeAddress())
+		if addr.Cmp(ss.localNodeAddress) == 0 {
+			st, err := ss.streamCache.GetStreamNoWait(ctx, streamID)
+			if err != nil {
+				return AsRiverError(err).Func("startSyncingByCookie")
+			}
+
+			v, err := st.GetViewIfLocal(ctx)
+			if err != nil {
+				return AsRiverError(err).Func("startSyncingByCookie")
+			}
+
+			sc, err = v.GetStreamSince(ctx, ss.localNodeAddress, cookie)
+			if err != nil {
+				return AsRiverError(err).Func("startSyncingByCookie")
+			}
+		} else {
+			client, err := ss.nodeRegistry.GetStreamServiceClientForAddress(addr)
+			if err != nil {
+				return AsRiverError(err).Func("startSyncingByCookie")
+			}
+
+			resp, err := client.GetStream(ctx, connect.NewRequest(&GetStreamRequest{
+				StreamId:   cookie.GetStreamId(),
+				SyncCookie: cookie,
+			}))
+			if err != nil {
+				return AsRiverError(err).Func("startSyncingByCookie")
+			}
+
+			sc = resp.Msg.GetStream()
+		}
+
+		// Send the stream since the given cookie to the subscription
+		if err := sub.messages.AddMessage(&SyncStreamsResponse{
+			SyncId:   syncID,
+			SyncOp:   SyncOp_SYNC_UPDATE,
+			Stream:   sc,
+			StreamId: cookie.GetStreamId(),
+		}); err != nil {
+			return AsRiverError(err).Func("startSyncingByCookie")
+		}
+	}
+
+	// Add the stream to the subscription
+	ss.router.addStreamToSubscription(syncID, streamID)
```

Comment:
Does the client miss an update when a new event was added in between `GetStreamSince` was called and the stream was added to the subscription?